### DISCUSSION TITLE
FeatureFloodCount on lower_dimensional_block (2D surface in 3D mesh)

### DISCUSSION THREAD

**Discussion by Thomas-Ulrich** [idaholab/moose#24879] [Q&A Modules: Phase field]
Hi,

I'm modeling fluid injection in a 3D geothermal system, including lower-dimensional features such as faults (2D surface and 1D well). I'm trying to check at each time step the areas of connected patches of faults that are overstressed (that is that comply with a given failure criterion).

To compute the failure criterion, I compute normal and shear reactions on the 2D fault based on the 3D stress and the fault normal orientation.
e.g. normal_traction = stress_ij * normal_i* normal_j  where normal is the fault local normal.

In my mesh, I have a sideset `Fault1` and an associated LowerDBlock `Fault1LD`, e.g.:
```
[Mesh]
  [./file]
    type = FileMeshGenerator
    file = ExtendedBottom_1703_small.msh
  [../]
  [./Fault1LDgenerator]
    type = LowerDBlockFromSidesetGenerator
    new_block_id = 111
    new_block_name = 'Fault1LD'
    sidesets = 'Fault1'
    input = file
  [../]
```

If I use FeatureFloodCount restricted on the sideset 'Fault1`, based on a variable defined on the whole domain (e.g. temperature T), then it works as expected (e.g. FeatureVolumeVectorPostprocessor returns a non zero area):
```
   [./flood_count]
     type = FeatureFloodCount
     variable = T
     # Must be turned on to build data structures necessary for FeatureVolumeVPP
     compute_var_to_feature_map = true
     threshold = 1e5
     execute_on = TIMESTEP_END
     boundary = 'Fault1'
   [../]
```
But what I need is to use FeatureFloodCount on the block Fault1LD, based on a variable only defined there.
For example:
```

[AuxVariables]
  [./stress_excess]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

  #my user defined aux kernel computing stress_excess only on the 2D elements
  [AuxKernels]
    [./stress_excess_ini_ker]
    type = FunctionAux
    variable = 'stress_excess'
    function = '-1e10'
    execute_on = 'initial'
  [../]

  [./stress_excess_ker]
    type = TigerStressExcessAux
    variable = 'stress_excess'
    total_stress = stress
    mu_s = 0.3
    cohesion = 0
    execute_on = 'TIMESTEP_END'
    block = 'Fault1LD Fault2LD'
  [../]
  [../]
  
   [./flood_count]
     type = FeatureFloodCount
     variable = stress_excess
     # Must be turned on to build data structures necessary for FeatureVolumeVPP
     compute_var_to_feature_map = true
     threshold = 1e5
     execute_on = TIMESTEP_END
   [../]
```
Then I get 0 in FeatureVolumeVectorPostprocessor, probably because LowerDBlocks are ignored by FeatureFloodCount?

How could I make FeatureFloodCount work on the block Fault1LD only? 
Do I need to modify the code FeatureFloodCount, or could I use existing moose elements to make it work?

Thanks in advance,

Thomas Ulrich.

**Comment by GiudGiud** [comment]
Hello

does it work if you boundary restrict the FeatureFloodCount? (Just a guess)

and did you confirm the auxkernel higher in the pipeline works as expected?

Guillaume

**Reply by Thomas-Ulrich** [reply]
Hi,
The FeatureFloodCount works if the boundary restricts the FeatureFloodCount when the variable is defined on the volume.
(in other words, this example modules/phase_field/test/tests/feature_volume_vpp_test/boundary_area_3D_single.i)
 works).
 
I think if I use the boundary restrictions (on FeatureFloodCount), FeatureFloodCount may use the solution on the volume elements containing the boundary, compute the solution on the boundary, and apply the FeatureFloodCount.

But my variable (stress excess) is already computed on the fault (on the lower dimensional fault, not the boundary), and not in the volume (well it is only initialized in the volume at the first step to a -1e10).

The auxkernel works fine (confirmed looking at the exodus file).
![image](https://github.com/idaholab/moose/assets/13536910/4d462138-75e6-42f1-bdb0-d75dd7eb2044)
![image](https://github.com/idaholab/moose/assets/13536910/dd8c546f-bd06-4f3b-8e99-4b3e07c2f77e)

**Reply by Thomas-Ulrich** [reply]
Hi,
So I m now able to restrict the FeatureFloodCount to the 2D LD elements only, with the following changes:
```diff
diff --git a/modules/phase_field/src/postprocessors/FeatureFloodCount.C b/modules/phase_field/src/postprocessors/FeatureFloodCount.C
index 10cd679cda..7c14feadc6 100644
--- a/modules/phase_field/src/postprocessors/FeatureFloodCount.C
+++ b/modules/phase_field/src/postprocessors/FeatureFloodCount.C
@@ -100,6 +100,7 @@ FeatureFloodCount::validParams()
 {
   InputParameters params = GeneralPostprocessor::validParams();
   params += BoundaryRestrictable::validParams();
+  params += BlockRestrictable::validParams();
 
   params.addRequiredCoupledVar(
       "variable",
@@ -191,6 +192,7 @@ FeatureFloodCount::FeatureFloodCount(const InputParameters & parameters)
     Coupleable(this, false),
     MooseVariableDependencyInterface(this),
     BoundaryRestrictable(this, false),
+    BlockRestrictable(this, false),
     _fe_vars(getCoupledMooseVars()),
     _vars(getCoupledStandardMooseVars()),
     _dof_map(_vars[0]->dofMap()),
@@ -372,6 +374,9 @@ FeatureFloodCount::execute()
   {
     for (const auto & current_elem : _mesh.getMesh().active_local_element_ptr_range())
     {
+
+      if (!hasBlocks(current_elem->subdomain_id()))
+        continue;
       // Loop over elements or nodes
       if (_is_elemental)
       {
(moose) ulrich@heisenbug:/export/dump/ulrich/myLibs/moose$ git diff modules/phase_field/include/postprocessors/FeatureFloodCount.h
diff --git a/modules/phase_field/include/postprocessors/FeatureFloodCount.h b/modules/phase_field/include/postprocessors/FeatureFloodCount.h
index e601a282ca..9f48fd5ece 100644
--- a/modules/phase_field/include/postprocessors/FeatureFloodCount.h
+++ b/modules/phase_field/include/postprocessors/FeatureFloodCount.h
@@ -14,6 +14,7 @@
 #include "InfixIterator.h"
 #include "MooseVariableDependencyInterface.h"
 #include "BoundaryRestrictable.h"
+#include "BlockRestrictable.h"
 
 #include <iterator>
 #include <list>
@@ -40,7 +41,8 @@ class MooseMesh;
 class FeatureFloodCount : public GeneralPostprocessor,
                           public Coupleable,
                           public MooseVariableDependencyInterface,
-                          public BoundaryRestrictable
+                          public BoundaryRestrictable,
+                          public BlockRestrictable
 {
 public:
   static InputParameters validParams();
```

I'm using the following postprocessors:
```
[Postprocessors]
  [./stress_excess_max]
    type = ElementExtremeValue
    variable = stress_excess
    execute_on = ' timestep_end'
  [../]
   [./flood_count]
     type = FeatureFloodCount
     variable = stress_excess
     # Must be turned on to build data structures necessary for FeatureVolumeVPP
     compute_var_to_feature_map = true
     threshold = 2.9e6
     execute_on = 'timestep_end'
     block = 'Fault1D Fault2D'
   [../]
  [./Volume]
    type = VolumePostprocessor
    execute_on = 'initial'
  [../]
  [./volume_fraction]
    type = FeatureVolumeFraction
    mesh_volume = Volume
    feature_volumes = feature_volumes
    execute_on = 'initial timestep_end'
  [../]
[]

[VectorPostprocessors]
  [./feature_volumes]
    type = FeatureVolumeVectorPostprocessor
    flood_counter = flood_count
    execute_on = 'initial timestep_end'
    output_centroids = True
    single_feature_per_element = False
    #outputs = none
  [../]
[]
```

But the areas computed by FeatureVolumeVectorPostprocessor are much larger than expected (e.g. when looking with paraview at the file with filter threshold).
I speculate that the wrong quadrature rule is used, or the wrong Jacobian when computing the integral here when using low-dimensional elements:
https://github.com/idaholab/moose/blob/next/modules/phase_field/src/vectorpostprocessors/FeatureVolumeVectorPostprocessor.C#L229
What do you think?

**Reply by GiudGiud** [reply]
@permcody you have worked on that feature right?

**Reply by Thomas-Ulrich** [reply]
In the end, if I use `single_feature_per_element = True` then the area gets realistic values, So I now have a working workflow.

**Reply by Thomas-Ulrich** [reply]
But my understanding is that the computeIntegral is not doing what it was intended for because it should not integrate the quantity (e.g. stress excess) but a derived quantity with value 1 if the quantity is above the threshold and 0 if below.
